tokio/src/sync/cancellation_token.rs

```diff
@@ -0,0 +1,1119 @@
+            inner: unsafe { NonNull::new_unchecked(Box::into_raw(state)) },
+        }
+    }
+
+    /// Returns a reference to the utilized `CancellationTokenState`.
+    fn state(&self) -> &CancellationTokenState {
+        // Safety: The state inside a `CancellationToken` is always valid, since
+        // is reference counted
+        unsafe { &*self.inner.as_ptr() }
+    }
+
+    /// Creates a `CancellationToken` which will get cancelled whenever the
+    /// current token gets cancelled.
+    ///
+    /// If the current token is already cancelled, the child token will get
+    /// returned in cancelled state.
+    ///
+    /// /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::CancellationToken;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let token = CancellationToken::new();
+    ///     let child_token = token.child_token();
+    ///
+    ///     let join_handle = tokio::spawn(async move {
+    ///         child_token.wait_for_cancellation().await;
+    ///         5
+    ///     });
+    ///
+    ///     tokio::spawn(async move {
+    ///         tokio::time::delay_for(std::time::Duration::from_millis(10)).await;
+    ///         token.cancel();
+    ///     });
+    ///
+    ///     assert_eq!(5, join_handle.await.unwrap());
+    /// }
+    /// ```
+    pub fn child_token(&self) -> CancellationToken {
+        let inner = self.state();
+
+        // Increment the refcount of this token. It will be referenced by the
+        // child, independent of whether the child is immediately cancelled or
+        // not.
+        let _current_state = inner.increment_refcount(inner.snapshot());
+
+        let mut unpacked_child_state = StateSnapshot {
+            has_parent_ref: true,
+            refcount: 1,
+            cancel_state: CancellationState::NotCancelled,
+        };
+        let mut child_token_state = Box::new(CancellationTokenState::new(
+            Some(self.inner),
+            unpacked_child_state,
+        ));
+
+        {
+            let mut guard = inner.synchronized.lock().unwrap();
+            if guard.is_cancelled {
+                // This task was already cancelled. In this case we should not
+                // insert the child into the list, since it would never get removed
+                // from the list.
+                (*child_token_state.synchronized.lock().unwrap()).is_cancelled = true;
+                unpacked_child_state.cancel_state = CancellationState::Cancelled;
+                // Since it's not in the list, the parent doesn't need to retain
+                // a reference to it.
+                unpacked_child_state.has_parent_ref = false;
+                child_token_state
+                    .state
+                    .store(unpacked_child_state.pack(), Ordering::SeqCst);
+            } else {
+                if let Some(first_child) = guard.first_child {
+                    child_token_state.from_parent.next_peer = Some(first_child);
```

Comment:
Needs to update the link in the other direction, i.e., `prev_peer` of `first_child`.

Reply:
Indeed! I think this was a left from when I started with a single linked list and then never looked again.

Reply:
I fixed that one now too and added a test for it. Loom still fails. But I would also have been surprised if it had made a difference for that one, since I had reduced the test to only a single child before, where the bug wouldn't have made a difference.